cli/pcluster/config/config_patch.py

```diff
@@ -0,0 +1,256 @@
+    output of a check will contain the following information:
+        - A boolean telling if the new configuration can be safely applied
+        - A list of change rows with all the information to build a detailed report
+    """
+
+    IGNORED_SECTIONS = ["global", "aliases"]  # Sections ignored for patch creation
+    FAIL_REASON_EMPTY = "-"  # Default value for empty fail reason
+
+    # Custom param comparators for special cases
+    PARAM_COMPARATORS = {}
+    PARAM_COMPARATORS["CLUSTER", "ebs_settings"] = compare_ebs_settings
+
+    def __init__(self, base_config, target_config):
+        """
+        Create a ConfigPatch.
+
+        :param base_config: The base configuration, f.i. as reconstructed from CloudFormation
+        :param target_config: The target configuration, f.i. as loaded from configuration file
+        """
+        # Cached condition results
+        self.condition_results = {}
+
+        self.base_config = copy.deepcopy(base_config)
+        self.target_config = copy.deepcopy(target_config)
+        self.changes = []
+        self._compare()
+
+    @property
+    def stack_name(self):
+        """Get the name of the stack this patch is referred to."""
+        return (
+            utils.get_stack_name(self.base_config.cluster_name) if hasattr(self.base_config, "cluster_name") else None
+        )
+
+    def _compare(self):
+        """
+        Compare target with base configuration.
+
+        All detected changes are added to the internal changes list, ready to be checked  through the public check()
+        method.
+        """
+        # Remove ignored sections
+        self._remove_ignored_sections(self.base_config)
+        self._remove_ignored_sections(self.target_config)
+
+        # First, compare all sections from target vs base config and mark visited base sections.
+        for section_key in sorted(self.target_config.get_section_keys()):
+            for section_label in sorted(self.target_config.get_sections(section_key).keys()):
+                target_section = self.target_config.get_section(section_key, section_label)
+                base_section = self._get_config_section(self.base_config, target_section)
+                base_section.visited = True
+                self._compare_section(base_section, target_section)
+
+        # Then, compare all non visited base sections vs target config.
+        for section_key in sorted(self.base_config.get_section_keys()):
+            for section_label in sorted(self.base_config.get_sections(section_key).keys()):
+                base_section = self.base_config.get_section(section_key, section_label)
+                if not hasattr(base_section, "visited"):
+                    target_section = self._get_config_section(self.target_config, base_section)
+                    self._compare_section(base_section, target_section)
+
+    def _compare_section(self, base_section, target_section):
```

Comment:
Eventually we could define a `compare` method for the Section object (see my next comment)

Reply:
I don't think it would simplify the code. 
The _compare_section method produces a set of Changes directly in the parent ConfigPatch. A compare method in the Section object would return a set of different objects which should be converted back into ConfigPatch.Change instances to be added to the patch.